src/decoder/cuda-decoder-utils.h

```diff
@@ -0,0 +1,418 @@
+
+// inline host device function definition
+
+union float_uint {
+  float f;
+  uint32 u;
+};
+inline HOST DEVICE uint32 float_as_uint(BaseFloat val) {
+  return ((float_uint*)&val)->u;
+}
+inline HOST DEVICE BaseFloat uint_as_float(uint32 val) {
+  return ((float_uint*)&val)->f;
+}
+
+// In atomic based token recombination, we pack the
+// cost and the arc index into an uint64 to represent the token
+// before recombination, with the former one in the higher bits
+// for comparison purpose.
+// for speedup purpose, make them inline (5% 0.165->0.158)
+inline HOST DEVICE uint64 pack_cost_idx_into_uint64(BaseFloat cost, int32 idx) {
+  uint32 i_cost = float_as_uint(cost);
+  if (i_cost & 0x80000000)
+    i_cost = i_cost ^ 0xFFFFFFFF;
+  else
+    i_cost = i_cost ^ 0x80000000;
+  return (uint64)i_cost << 32 | idx;
+}
+
+// Unpacks a cost
+inline HOST DEVICE BaseFloat unpack_cost_from_uint64(uint64 packed) {
+  uint32 i_cost = packed >> 32;
+  if (i_cost & 0x80000000)
+    i_cost = i_cost ^ 0x80000000;
+  else
+    i_cost = i_cost ^ 0xFFFFFFFF;
+  return uint_as_float(i_cost);
+}
+
+// Unpacks a idx for tracing the data
+inline HOST DEVICE int32 unpack_idx_from_uint64(uint64 packed) {
+  return packed & 0x7FFFFFFF;
+}
+
+// inline device function definition
+#ifdef __CUDACC__
+
+// another choice is pre-store the result of this binsearch, e.g. do something like:
+// for (int j=0; j < params->d_degrees[idx]; j++)
+//  params->d_lowerbound[j + params->d_degrees_scan[idx]] = idx;
+// however this for loop costs 9% more time, while the binsearch costs 6% time
+// thus the fastest way is doing binsearch without pre-store
+inline DEVICE int binsearch_maxle(const int *vec, const int val, int low,
+                                  int high) {
+  while (true) {
+    if (low == high)
+      return low; //we know it exists
+    if ((low + 1) == high)
+      return (vec[high] <= val) ? high : low;
+
+    int mid = low + (high - low) / 2;
+
+    if (vec[mid] > val)
+      high = mid - 1;
+    else
+      low = mid;
+  }
+}
+
+
+// fast load 16 bytes using CUDA ASM
+inline  DEVICE void fast_load16(void *a, const void *b) {
+  const ulong2 *src = reinterpret_cast<const ulong2*>(b);
+  ulong2 &dst = *reinterpret_cast<ulong2*>(a);
+  asm("ld.global.v2.u64 {%0,%1}, [%2];" : "=l"(dst.x), "=l"(dst.y) : "l"(src));
+}
+
+// fast store 16 bytes using CUDA ASM
+inline  DEVICE void fast_store16(void *a, const void *b) {
+  const ulong2 src = *reinterpret_cast<const ulong2*>(b);
+  asm("st.global.v2.u64 [%0], {%1,%2};" :: "l"(a), "l"(src.x), "l"(src.y));
+}
+
+// fast store 8 bytes using CUDA ASM
+inline  DEVICE void fast_store8(void *a, const void *b) {
+  *(uint64*)a = (*(uint64*)b);
+}
+
+// TODO: we need a fast 32 bytes storing function
+inline  DEVICE void fast_store32(void *a, const void *b) {
+  memcpy(a, b, 32);
+}
+
+// overload CUDA atomicMin to consume double
+inline DEVICE double atomic_min(double *address, double val) {
```

Comment:
Hi @chenzhehuai, why are these two atomic_min definitions here? I see nothing wrong with simply calling atomicMin by itself. In fact, I am certain that that will be faster (and less confusing for people reading the code).

Reply:
Yes, I agree. We should try your proposal.